apiserver/facades/client/storage/storage.go

```diff
@@ -306,19 +345,166 @@ func (a *StorageAPI) Remove(ctx context.Context, args params.RemoveStorage) (par
+		result = append(result, params.ErrorResult{
+			Error: apiservererrors.ServerError(err),
+		})
+	}
 
 	return params.ErrorResults{Results: result}, nil
 }
 
+// detachStorageAttachment detaches exactly one storage attachment for a
+// request. It expects that the caller has processed the supplied tags and can
+// now provide string values representing the entities.
+func (a *StorageAPI) detachStorageAttachment(
+	ctx context.Context,
+	storageID string,
+	unitName coreunit.Name,
+	force bool,
+	wait time.Duration,
+) error {
+	unitUUID, err := a.applicationService.GetUnitUUID(ctx, unitName)
+	switch {
+	case errors.Is(err, coreunit.InvalidUnitName):
+		return errors.Errorf("invalid unit name %q", unitName).Add(
+			coreerrors.NotValid,
+		)
+	case errors.Is(err, applicationerrors.UnitNotFound):
+		return errors.Errorf("unit %q does not exist", unitName).Add(coreerrors.NotFound)
+	case err != nil:
+		return errors.Errorf(
+			"getting unit uuid for unit name %q: %w", unitName, err,
+		)
+	}
+
+	storageInstanceUUID, err := a.storageService.GetStorageInstanceUUIDForID(
+		ctx, storageID,
+	)
+	switch {
+	case errors.Is(err, storageerrors.StorageInstanceNotFound):
+		return errors.Errorf("storage %q does not exist", storageID).Add(
+			coreerrors.NotFound,
+		)
+	case err != nil:
+		return errors.Errorf(
+			"getting storage instance uuid for storage id %q: %w", storageID, err,
+		)
+	}
+
+	storageAttachmentUUID, err := a.storageService.
+		GetStorageAttachmentUUIDForStorageInstanceAndUnit(
+			ctx, storageInstanceUUID, unitUUID,
+		)
+	// We purposely ignore not valid errors for the uuids supplied. We have
+	// recieved these uuids from the domain and not the caller so they can
+	// safely be considered valid.
+	switch {
+	case errors.Is(err, storageerrors.StorageInstanceNotFound):
+		return errors.Errorf("storage %q does not exist", storageID).Add(coreerrors.NotFound)
+	case errors.Is(err, storageerrors.StorageAttachmentNotFound):
+		return errors.Errorf(
+			"storage %q is not attached to unit %q", storageID, unitName,
+		).Add(coreerrors.NotValid)
+	case errors.Is(err, applicationerrors.UnitNotFound):
+		return errors.Errorf("unit %q does not exist", unitName).Add(coreerrors.NotFound)
+	case err != nil:
+		return errors.Errorf(
+			"getting storage attachment uuid for storage %q attached to unit %q: %w",
+			storageID, unitName, err,
+		)
+	}
+
+	_, err = a.removalService.RemoveStorageAttachmentFromAliveUnit(
+		ctx, storageAttachmentUUID, force, wait,
+	)
+	// Early exit path. Everything below this point can now be considered error
+	// handling.
+	if err == nil {
+		return nil
+	}
+
+	viErr, has := errors.AsType[applicationerrors.UnitStorageMinViolation](err)
+	if has {
+		return errors.Errorf(
+			"removing storage %q from unit %q would violate charm storage %q requirements of having minimum %d storage instances",
+			storageID, unitName, viErr.CharmStorageName, viErr.RequiredMinimum,
+		)
+	}
```

Comment:
I really hate this. I regret us doing it this way.

In hindsight, breaking apart assertion and casting from each other results in a nicer API.
It does force you to define two errors, but is it so much cleaner to use in both `if` and `switch` blocks.

```go
if errors.Is(err, applicationerrors.UnitStorageMinViolation) {
    v := errors.AsType[applicationerrors.UnitStorageMinViolationDetails](err)
    //...
}

switch {
    case errors.Is(err, applicationerrors.UnitStorageMinViolation):
        v := errors.AsType[applicationerrors.UnitStorageMinViolationDetails](err)
        //...
}
```